# Assignment 3 - Text Analysis
An explanation this assignment could be found in the .pdf explanation document


## Materials to review for this assignment
<h4>From Moodle:</h4> 
<h5><u>Review the notebooks regarding the following python topics</u>:</h5>
<div class="alert alert-info">
&#x2714; <b>Working with strings</b> (tutorial notebook)<br/>
&#x2714; <b>Text Analysis</b> (tutorial notebook)<br/>
&#x2714; <b>Hebrew text analysis tools (tokenizer, wordnet)</b> (moodle example)<br/>
&#x2714; <b>(brief review) All previous notebooks</b><br/>
</div> 
<h5><u>Review the presentations regarding the following topics</u>:</h5>
<div class="alert alert-info">
&#x2714; <b>Text Analysis</b> (lecture presentation)<br/>
&#x2714; <b>(brief review) All other presentations</b><br/>
</div>

## Personal Details:

In [1]:
# Details Student 1:

# Details Student 2:


## Preceding Step - import modules (packages)
This step is necessary in order to use external modules (packages). <br/>

In [3]:
# --------------------------------------
import pandas as pd
import numpy as np
# --------------------------------------


# --------------------------------------
# ------------- visualizations:
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
# --------------------------------------


# ---------------------------------------
import sklearn
from sklearn import preprocessing, metrics, pipeline, model_selection, feature_extraction 
from sklearn import naive_bayes, linear_model, svm, neural_network, neighbors, tree
from sklearn import decomposition, cluster

from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV 
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.metrics import precision_score, recall_score, f1_score
from sklearn.metrics import mean_squared_error, r2_score, silhouette_score
from sklearn.preprocessing import MinMaxScaler, StandardScaler, LabelEncoder

from sklearn.svm import LinearSVC
from sklearn.neural_network import MLPClassifier
from sklearn.linear_model import Perceptron, SGDClassifier
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.naive_bayes import MultinomialNB, GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
# ---------------------------------------


# ----------------- output and visualizations: 
import warnings
from sklearn.exceptions import ConvergenceWarning
warnings.simplefilter("ignore")
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter("ignore", category=ConvergenceWarning)
# show several prints in one cell. This will allow us to condence every trick in one cell.
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
%matplotlib inline
pd.pandas.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%.3f' % x)
# ---------------------------------------

### Text analysis and String manipulation imports:

In [8]:
# --------------------------------------
# --------- Text analysis and Hebrew text analysis imports:
# vectorizers:
from sklearn.feature_extraction import text
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

# regular expressions:
import re
# --------------------------------------

### (optional) Hebrew text analysis - WordNet (for Hebrew)
Note: the WordNet is not a must

#### (optional) Only if you didn't install Wordnet (for Hebrew) use:

In [4]:
# word net installation:

# unmark if you want to use and need to install
# !pip install wn
# !python -m wn download omw-he:1.4

In [5]:
# word net import:

# unmark if you want to use:
# import wn

### (optional) Hebrew text analysis - hebrew_tokenizer (Tokenizer for Hebrew)
Note: the hebrew_tokenizer is not a must

#### (optional) Only if you didn't install hebrew_tokenizer use:

In [6]:
# Hebrew tokenizer installation:

# unmark if you want to use and need to install:
#!pip install hebrew_tokenizer

In [7]:
# Hebrew tokenizer import:

# unmark if you want to use:
#import hebrew_tokenizer as ht

### Reading input files
Reading input files for train annotated corpus (raw text data) corpus and for the test corpus

In [4]:
train_filename = 'annotated_corpus_for_train.csv'
test_filename  = 'corpus_for_test.csv'
df_train = pd.read_csv(train_filename, index_col=None, encoding='utf-8')
df_test  = pd.read_csv(test_filename, index_col=None, encoding='utf-8')

In [5]:
df_train.head(8)
df_train.shape

story gender
0  כשחבר הזמין אותי לחול, לא באמת חשבתי שזה יקרה,...      m
1  לפני שהתגייסתי לצבא עשיתי כל מני מיונים ליחידו...      m
2  מאז שהתחילו הלימודים חלומו של כל סטודנט זה הפנ...      f
3  כשהייתי ילד, מטוסים היה הדבר שהכי ריתק אותי. ב...      m
4  ‏הייתי מדריכה בכפר נוער ומתאם הכפר היינו צריכי...      f
5  לפני כ3 חודשים טסתי לרומא למשך שבוע. טסתי במטו...      f
6  אני כבר שנתיים נשוי והשנה אני ואישתי סוף סוף י...      m
7  השנה התחלנו שיפוץ בדירה שלנו בתל אביב. הדירה ה...      f

(753, 2)

In [6]:
df_test.head(3)
df_test.shape

test_example_id                                              story
0                0  כל קיץ אני והמשפחה נוסעים לארצות הברית לוס אנג...
1                1  הגעתי לשירות המדינה אחרי שנתיים כפעיל בתנועת "...
2                2  אחת האהבות הגדולות שלי אלו הכלבים שלי ושל אישת...

(323, 2)

### Your implementation:
Write your code solution in the following code-cells

# Preprocessing

The preprocessing part of the text involves replacing certain characters at word endings, removing specific word prefixes and combinations, filtering out short words, and removing punctuation. This prepares the text for analysis by splitting it into words, removing unwanted characters, and ensuring meaningful word representations for further processing.


In [15]:
def custom_preprocessing(text):
    # Replace characters
    text = re.sub(r'מ\b', 'ם', text)  # Replace 'מ' at word endings with 'ם'
    text = re.sub(r'נ\b', 'ן', text)  # Replace 'נ' at word endings with 'ן'
    text = re.sub(r'פ\b', 'ף', text)  # Replace 'פ' at word endings with 'ף'
    text = re.sub(r'כ\b', 'ך', text)  # Replace 'כ' at word endings with 'ך'
    text = re.sub(r'צ\b', 'ץ', text)  # Replace 'צ' at word endings with 'ץ'
    
    text = re.sub(r'\bכש', '', text)  # Remove 'כש' from start of the word
    text = re.sub(r'\bלכש', '', text)  # Remove 'לכש' from start of the word
    text = re.sub(r'\bוש', '', text)  # Remove 'וש' from start of the word
    text = re.sub(r'\bב', '', text)   #Remove 'ב' from start of the word 
    text = re.sub(r'\bו', '', text)   #Remove 'ו' from start of the word
    
    words = text.split()
    #Kepp words only with length > 2 
    words = [word for word in words if len(word) > 2]
    
    # Define the punctuation characters you want to remove
    punctuation_chars = r'[.,?!$%@*()\[\]{}"\'<>:/\\;-]'

    # Create a translation table that maps the punctuation characters to None
    translation_table = str.maketrans(punctuation_chars, ' '*len(punctuation_chars))

    # Apply the translation table to each word to remove punctuation characters
    words = [word.translate(translation_table) for word in words]
    
    # Join the preprocessed words back into a string
    processed_text = ' '.join(words)
    
    return processed_text
    

# Apply custom preprocessing to train and test texts
df_train['story'] = df_train['story'].apply(custom_preprocessing)
df_test['story'] = df_test['story'].apply(custom_preprocessing)

# Print the preprocessed text of the second entry in the training dataset
print(df_train['story'][1])




לפני שהתגייסתי לצבא עשיתי מני מיונים ליחידות מובחרות יום סיירות אותו יום אכן קיבלתי מיון ליחידה מאד מאד מיוחדת חיל הים אני כאחד שלא אוהב מיוחד מים אמרתי שאתן לזה סיכוי אנסה ללכת לראות יצא מהנסיון המעניין הזה הגענו לאיזה מקום אזור חיפה לאחר כמה דקות לבד אמרו לנו ישר להוריד הדברים ישר התחילו להריץ אותנו ממקום למקום מלמעלה למטה זחילות ריצה דרך שאפשר לדמיין השלבים הקשיים יותר היו הריצות למים לשחות שעות ארוכות להגיע ממקום למקום קבוצה יחידים אוהל שלי היו עוד שתי אנשים שאחד היה מאד מורעל להתקבל יהיר דברן רעשן הבן אדם השני היה אדם שקט נחמד מופנם שהגענו לאתר ידענו לכמה זמן אנחנו מגיעים למיונים עושים איך עושים ידענו שאנחשים הולכים הביתה תוך כדי השמועות אמרו שהמיון אמור להיות ימים לשבוע היו המון אנשים התחלה לאט לאט התחלנו להרגיש הדילול אנשים מרגע לרגע היה קיצוני היינו עושים תחרות סמויה האוהלים מבין האוהלים נשארים הכי הרבה חברי צוות אוהל היה שלושה אנשים יום אחרי יום אנשים עפים ימין שמאל אמת ראינו שאנחנו האוהל שנשאר הרכב המלא שלו התחלנו לדבר אוהל הבחור היהיר התחיל להתרברב לצחוק כולם הבחור השני אמר 

# Vectorization

The vectorization stage converts preprocessed text into numerical format using TF-IDF. Custom tokenizer splits text into words. X_train_tfidf and X_test_tfidf hold transformed features. IDF scores quantify word importance, seen in top and bottom word lists.

In [10]:
# Tokenization
def custom_tokenizer(text):
    return text.split()

# Define labels
y_train = df_train['gender']

# Vectorize training data
vectorizer = TfidfVectorizer(tokenizer=custom_tokenizer, max_features=10000)
X_train= vectorizer.fit_transform(df_train['story'])

# Vectorize test data
X_test = vectorizer.transform(df_test['story'])

# Print the top and bottom words with IDF scores
feature_names = vectorizer.get_feature_names_out()
idf_scores = vectorizer.idf_
word_idf_dict = {feature_names[i]: idf_scores[i] for i in range(len(feature_names))}
sorted_words = sorted(word_idf_dict.items(), key=lambda x: x[1], reverse=True)

print("Top 10 words with highest IDF scores:")
for word, idf in sorted_words[:10]:
    print(f"{word}: {idf:.4f}")
print()

print("Last 10 words with lowest IDF scores:")
for word, idf in sorted_words[-10:]:
    print(f"{word}: {idf:.4f}")
print()

Top 10 words with highest IDF scores:
10k: 6.9322
669: 6.9322
i25: 6.9322
mall: 6.9322
piccolo: 6.9322
wolt: 6.9322
אבבה: 6.9322
אדריכלים: 6.9322
אובט: 6.9322
אודישן: 6.9322

Last 10 words with lowest IDF scores:
הייתה: 1.5184
יום: 1.5095
הוא: 1.4985
מאוד: 1.4942
אבל: 1.4877
אחד: 1.4663
לאחר: 1.4289
שלי: 1.2903
אני: 1.1686
היה: 1.1241



# Machine Learning and Evaluation

In this phase, we're unleashing the potential of top classifiers: Naive Bayes, Decision Tree, and K-Nearest Neighbors, chosen for their diverse strengths. Each model is integrated into a pipeline, simplifying the process. By testing these models using cross-validation on preprocessed text data, we're striving for optimal performance, as measured by the F1 score. Customized hyperparameters for each model are revealed, contributing to their accuracy. The average F1 scores give us a clear view of their effectiveness in gender prediction.

In [11]:
# Hyperparameter tuning for Naive Bayes
nb_model = GridSearchCV(MultinomialNB(),
                        {'alpha': [0.01, 0.1, 0.07, 0.06, 0.05, 0.5, 0.04, 1.0], 'fit_prior': [True, False]},
                        scoring='f1_macro', cv=10)
nb_model.fit(X_train, y_train)

# Hyperparameter tuning for Decision Tree
dt_model = GridSearchCV(DecisionTreeClassifier(),
                        {'criterion': ['gini', 'entropy'], 'max_depth': [5, 10, 20, 25, 30], 'min_samples_split': [2, 5, 10]},
                        scoring='f1_macro', cv=10)
dt_model.fit(X_train, y_train)

# Hyperparameter tuning for K-Nearest Neighbors
knn_model = GridSearchCV(KNeighborsClassifier(),
                         {'n_neighbors': [3, 5, 7, 9], 'weights': ['uniform', 'distance']},
                         scoring='f1_macro', cv=10)
knn_model.fit(X_train, y_train)

# Print best hyperparameters and average F1 scores
print("Best Hyperparameters - Naive Bayes Classifier:")
print(nb_model.best_params_)
print()
print("Average F1 Score - Naive Bayes Classifier:")
print(f"{nb_model.best_score_:.4f}")
print()
print("Best Hyperparameters - Decision Tree Classifier:")
print(dt_model.best_params_)
print()
print("Average F1 Score - Decision Tree Classifier:")
print(f"{dt_model.best_score_:.4f}")
print()
print("Best Hyperparameters - K-Nearest Neighbors Classifier:")
print(knn_model.best_params_)
print("Average F1 Score - K-Nearest Neighbors Classifier:")
print(f"{knn_model.best_score_:.4f}")

GridSearchCV(cv=10, estimator=MultinomialNB(),
             param_grid={'alpha': [0.01, 0.1, 0.07, 0.06, 0.05, 0.5, 0.04, 1.0],
                         'fit_prior': [True, False]},
             scoring='f1_macro')

GridSearchCV(cv=10, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [5, 10, 20, 25, 30],
                         'min_samples_split': [2, 5, 10]},
             scoring='f1_macro')

GridSearchCV(cv=10, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [3, 5, 7, 9],
                         'weights': ['uniform', 'distance']},
             scoring='f1_macro')

Best Hyperparameters - Naive Bayes Classifier:
{'alpha': 0.05, 'fit_prior': False}

Average F1 Score - Naive Bayes Classifier:
0.6450

Best Hyperparameters - Decision Tree Classifier:
{'criterion': 'gini', 'max_depth': 20, 'min_samples_split': 10}

Average F1 Score - Decision Tree Classifier:
0.6186

Best Hyperparameters - K-Nearest Neighbors Classifier:
{'n_neighbors': 5, 'weights': 'distance'}
Average F1 Score - K-Nearest Neighbors Classifier:
0.6038


# Results

The Naive Bayes classifier with hyperparameters (alpha: 0.05, fit_prior: False) demonstrated the highest average F1 score of 0.6450, indicating robust performance in balancing precision and recall. Although the Decision Tree and K-Nearest Neighbors  classifiers also performed well with average F1 scores of 0.6186 and 0.6038 respectively, the Naive Bayes classifier exhibited the strongest predictive capabilities for the given task.

### Save output to csv (optional)
After you're done save your output to the 'classification_results.csv' csv file.<br/>
We assume that the dataframe with your results contain the following columns:
* column 1 (left column): 'test_example_id'  - the same id associated to each of the test stories to be predicted.
* column 2 (right column): 'predicted_category' - the predicted gender value for each of the associated story. 

Assuming your predicted values are in the `df_predicted` dataframe, you should save you're results as following:

In [13]:
# Create a copy of df_test
predicted_df = df_test.copy()

# Add the predicted gender column to predicted_df
y_pred = nb_model.predict(X_test)
predicted_df['predicted_gender'] = y_pred


# Save the predictions to a CSV file
predicted_df.to_csv('classification_results.csv', index=False)
predicted_df



test_example_id                                              story   
0                  0  קיץ אני המשפחה נוסעים לארצות הברית לוס אנג לס ...  \
1                  1  הגעתי לשירות המדינה אחרי שנתיים כפעיל תנועת  י...   
2                  2  אחת האהבות הגדולות שלי אלו הכלבים שלי אישתי  ל...   
3                  3  רגע הגיוס לצבא היה הרגע הכי משמעותי עבורי  אני...   
4                  4  אני הגעתי לברזיל ישר מקולומביה אני עשיתי עבודת...   
..               ...                                                ...   
318              318  שנה האחרונה הרגשתי תקוע עבודה  השגרה הפכה למשה...   
319              319  אני אילן חברים טובים מזה שנה תמיד חלמנו לפתח ר...   
320              320  מידי יום שישי אני נוהג לנסוע תחבורה ציבורית סת...   
321              321  לפני מספר חודשים  שיא התחלואה הגל השני  עמית ש...   
322              322  היום דיווחתי גניבה האוטו שלי  אחד התעוררתי איח...   

    predicted_gender  
0                  m  
1                  m  
2                  m  
3                  m  
4                  m  
..               ...  
318                m  
319                m  
320                m  
321                m  
322                m  

[323 rows x 3 columns]